### ISSUE TITLE
typealias bug?

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#2552]
``` julia
julia> typealias SelfDict{T} Dict{T,T}
Dict{T,T}

julia> isself(d::Dict) = false
# methods for generic function isself
isself(d::Dict{K,V}) at none:1

julia> isself(d::SelfDict) = true
# methods for generic function isself
isself(d::Dict{T,T}) at none:1

julia> isself([1=>2])
true

julia> isself([1=>"foo"])
true
```

**Comment by mbauman** [comment]
Is this the same root cause as the following method ordering bug?

```jl
julia> type A{T<:Integer}; end

julia> f{T}(::A{T}) = 1 # Julia doesn't consider that T must be <: Integer here
f (generic function with 1 method)

julia> f(::A, ::Any...) = 2 # But does so here
f (generic function with 2 methods)

julia> f(A{Int}())  # So method 2 is more specific
2

julia> g{T<:Integer}(::A{T}) = 1 # But if I manually tell Julia about the restriction
g (generic function with 1 method)

julia> g(::A, ::Any...) = 2
g (generic function with 2 methods)

julia> g(A{Int}())  # Then dispatch works as expected
1
```

**Comment by yuyichao** [comment]
@mbauman Feels like a variant of https://github.com/JuliaLang/julia/issues/11597 ?

**Comment by mbauman** [comment]
Yes, that's much more applicable.  I came here from #7453, which also seemed similar but was closed as a duplicate of this guy.  It's not really a typealias problem though — it's a restricted type parameter.  So I wasn't sure if the same restricting/unrestricting problem was at the root cause here.

**Comment by yuyichao** [comment]
Actually thanks for pinging this, since I actually notice exactly this a few days ago.

**Comment by vtjnash** [comment]
removing the milestone tag since this is being tracked through https://github.com/JuliaLang/julia/issues/8974

**Comment by JeffBezanson** [comment]
closed by #18457
